### ISSUE TITLE
LFS Client Error (InvalidArgumentError)

### ISSUE THREAD

**Issue by rstecca** [git-lfs/git-lfs#3494]
Hello,
trying to perform a git lfs fetch but it just hangs (0 B/s for a while) and then quit with 

`
LFS: Client error: https://api.media.atlassian.com/file/992a6467-031c-499d-841f-19711e45640d/binary
`

The command
 `
GIT_TRACE=1 GIT_CURL_VERBOSE=1 git lfs fetch
`
results in a long series of errors like this one below:

`
< HTTP/1.1 400 Bad Request                                                                              < Connection: close                                                                                     < Content-Length: 167                                                                                   < Access-Control-Allow-Origin: *                                                                        < Access-Control-Expose-Headers: Accept-Ranges, Content-Encoding, Content-Length, Content-Range         < Content-Type: application/json                                                                        < Date: Thu, 24 Jan 2019 15:14:19 GMT                                                                   < Strict-Transport-Security: max-age=15552000; includeSubDomains                                        < X-B3-Spanid: 13f3b59a8f95a973                                                                         < X-B3-Traceid: 13f3b59a8f95a973                                                                        < X-Content-Type-Options: nosniff                                                                       < X-Dns-Prefetch-Control: off                                                                           < X-Download-Options: noopen                                                                            < X-Frame-Options: SAMEORIGIN                                                                           < X-Xss-Protection: 1; mode=block                                                                       <                                                                                                       15:14:22.479535 trace git-lfs: HTTP: {"error":{"code":"InvalidArgumentError","title":"range first index  must not be greater than last index","href":"https://api.media.atlassian.com#InvalidArgumentError"}}   {"error":{"code":"InvalidArgumentError","title":"range first index must not be greater than last index" ,"href":"https://api.media.atlassian.com#InvalidArgumentError"}}15:14:22.479535 trace git-lfs: tq: retr ying object 77dfe7e27990ba66880d64d5423e862b37cb7461405cdc03789e3bf055a050e1: LFS: Client error: https: //api.media.atlassian.com/file/7392df22-bf24-4717-aa21-7efb92efa827/binary                              15:14:22.479535 trace git-lfs: tq: enqueue retry #1 for "77dfe7e27990ba66880d64d5423e862b37cb7461405cdc 03789e3bf055a050e1" (size: 122880044)                                                                   15:14:22.588244 trace git-lfs: filepathfilter: rewrite ".git" as "**/.git/**"                           15:14:22.588244 trace git-lfs: filepathfilter: rewrite "**/.git" as "**/.git"                           15:14:22.588244 trace git-lfs: filepathfilter: rejecting "tmp" via []                                   15:14:22.588244 trace git-lfs: filepathfilter: accepting "tmp"
`

I'd put this in evidence as should be relevant:
`
trace git-lfs: HTTP: {"error":{"code":"InvalidArgumentError","title":"range first index  must not be greater than last index","href":"https://api.media.atlassian.com#InvalidArgumentError"}
`

I suspected the service was down but all operational on the bitbucket status page.
What am I doing wrong?

**Comment by bk2204** [comment]
Hey, sorry you've having trouble.

It sounds like what's happened is that one or more of the files you're trying to download are partially downloaded, and for some reason they have more data than they're supposed to. When we download a file, we send a `Range` header if the file already exists, and the server has complained that the low end of the range is larger than the high end (which is of course nonsensical).

I have a branch on my remote called `retry-on-oversize-file`. Can you try it and see if it fixes your problem?

**Comment by Apo-** [comment]
@rstecca I've been having the same issue since yesterday (repo is also hosted on **bitbucket**). 

I got this issue when CI was pulling, so then I tried cloning the repo locally, but then I get `Encountered <n> file(s) that should have been pointers, but weren't`. The `git lfs migrate import` and `git lfs uninstall && git reset --hard` didn't work. If I try to `lfs pull` anyway I get the same error and also `Unsolicited response received on idle HTTP channel starting with`.

@bk2204 I don't get what you meant by 
> I have a branch on my remote called retry-on-oversize-file. Can you try it and see if it fixes your problem?

**Comment by bk2204** [comment]
@Apo- I have a commit on the branch at https://github.com/bk2204/git-lfs/tree/retry-on-oversize-file which I think will address this bug. If possible, it would be nice if you or @rstecca could compile that branch and see if it fixes the `InvalidArgumentError` errors that you're seeing. (It won't fix any other errors that you're seeing, though.)

**Comment by Apo-** [comment]
@bk2204 It didn't do the trick for me, but I'm posting the process I followed in case it solves @rstecca problem (for Ubuntu 18.04):

```bash
# Install go
# Based on https://www.digitalocean.com/community/tutorials/how-to-install-go-on-ubuntu-18-04
cd ~
curl -O https://dl.google.com/go/go1.11.5.linux-amd64.tar.gz
tar xvf go1.11.5.linux-amd64.tar.gz
sudo chown -R root:root ./go
sudo mv go /usr/local
sudo nvim ~/.profile
# export GOPATH=$HOME/work
# export PATH=$PATH:/usr/local/go/bin:$GOPATH/bin
source  ~/.profile
mkdir $HOME/work

# Build git-lfs
go get github.com/bk2204/git-lfs
cd src/github.com/bk2204/git-lfs/
git checkout retry-on-oversize-file
nvim Makefile
# Replaced lint target with
# .PHONY : lint
# lint :

# Replace git-lfs
git lfs uninstall
sudo mv /usr/bin/git-lfs /usr/bin/git-lfs.old
sudo mv ~/work/bin/git-lfs /usr/bin
sudo chown root:root /usr/bin/git-lfs
git lfs install
```

**Comment by xander-sereda** [comment]
I've encountered this exact error - also on Bitbucket with git LFS. I've tried everything I can think of. Increased LFS timeouts, disabled smudge, nothing works.

In my case it's occurring for 34 different files.

**Comment by fraserr** [comment]
Having the same issue also.

**Comment by Apo-** [comment]
I opened an issue on bitbucket as well in case this is not a git-lfs issue https://bitbucket.org/site/master/issues/18064/git-lfs-fetch-fails-with

**Comment by Rustygames** [comment]
I contacted bitbucket support and they've fixed the issue for me! Here is their response:

```
We have been able to reproduce this issue on our end. We have identified that this issue is triggered by slow network connections at the time of pulling down the LFS files. This leads to a timeout on our backend.

I would like to inform that our team have identified the issue and have deployed a fix in regards to this.
```

**Comment by bbain** [comment]
TL;DR: Trying remove the files in your `.git/lfs/incomplete` directory and trying again.

I think I have tracked down what is happening here. There was an outage at Atlassian where a % of requests bad corrupted downloads of the correct size.

When the `git-lfs` is downloading a file its placed into `.git/lfs/incomplete`. When `git-lfs` got enough data (i.e. the file is of the correct size) it detected that the download was bad and returned an error. The problem is that the temporary file of correct size in `.git/lfs/incomplete` is still there.  

From this point on `git lfs fetch` or whatever will fail. The `git-lfs` sees the temporary file and incorrectly assumes that more data is required. It issues a HTTP request with `Range` header where starting from the temporary filesize + 1. Well, this is invalid because filesize + 1 is beyond the EOF.

I would assume that:

1. `git-lfs` shouldn't try and do a partial download when all the temporary file is the right size (but maybe wrong content).
2. `git-lfs` probably should delete temporary files that it detects as corrupted and just try again. Maybe this happens for free if the step 1 is implemented. This could introduce its own loop I guess. 

This is pretty easy to fake:

1. Create a repo with one LFS file >100MiB (to make things easy).
2. Start a clone but interrupt it before it completes.
3. Do a `cat /dev/urandom > .git/lfs/incomplete/<file>.tmp`. Kill before the tmp file reaches the size of the LFS file.
4. Do a `git lfs pull`. Make sure the upstream supports partial downloads. This will fail with an error.
5. Do a `git lfs pull`. This will issue HTTP requests with Range past the end of the file.

**Comment by bk2204** [comment]
I think it's certainly fair that we retry if we get the wrong amount of data or would issue an invalid Range request. I believe the proposed branch does that.

However, I'm not sure retrying when our download has data and it's corrupt is a good idea, since generally our data has come over an HTTPS connection, which verifies its integrity. That's a situation that we would want to produce an error for, which we do. In fact, doing so probably led to Bitbucket noticing that there was a problem.

**Comment by bbain** [comment]
> However, I'm not sure retrying when our download has data and it's corrupt is a good idea, since generally our data has come over an HTTPS connection, which verifies its integrity.

OK. The main problem is that the `git-lfs` will end up fail loop that a user wont know how to fix. Maybe a message telling users that something is very wrong and here's how to retry (e.g. by deleting this file)?

**Comment by bk2204** [comment]
I think it's fine to delete the file and retry, which is what my branch does in this case. I plan to spend a few minutes verifying this case and writing up some tests, and then submitting it.

**Comment by rstecca** [comment]
Thank you all for the responses. This is all very insightful. I also found that the combination Git-lfs + bitbucket + ssh is very problematic in itself and it causes misterious failures (even now that @Rustygames had bitbucket people's support), cul-de-sacs and cryptic errors. It can be difficult to understand what's failing. Recloning through https seems to be the way to avoid some of the snake pits. Not a solution but at least there's one way to not waste hours, days or weeks until ssh support is more solid. I want to switch back to ssh as soon as possible so I'll keep checking here and try @bk2204 's tricks.
@bbain amazing investigation. Good to keep in mind for future failures.

**Comment by rstecca** [comment]
@bk2204 

> 
> 
> I think it's fine to delete the file and retry, which is what my branch does in this case. I plan to spend a few minutes verifying this case and writing up some tests, and then submitting it.

I'm very curious to see what's the minimal failure case here. Can you post it?
>
> That's a situation that we would want to produce an error for, which we do. In fact, doing so probably led to Bitbucket noticing that there was a problem

Oh yes please. A meaningful error for this could have saved a lot of time for me. Cool that this post and your efforts escalated to bitbucket!

**Comment by bk2204** [comment]
I've opened #3518 which has a test for this case, which should demonstrate the problem. My hope is that it will make it into Git LFS 2.7.0.
